### PR TITLE
Iss797 parse intem emissions

### LINKED ISSUE #797: Read and interpret InTEM format (emissions data)
### Is your feature request related to a problem?

We need to be able to understand and read InTEM output data into OpenGHG. This will be both:

 - 1D emissions totals (and uncertainties)
 - 2D + time emissions maps (no uncertainties)

We should confirm with Alistair Manning about the latest output format for this.

### Describe the solution you'd like

_No response_

### Additional context

This will include 1D emissions data and so Issue #301 will need to be completed for this.

### PR DESCRIPTION
**Pull request by SutarPrasad**
* **Summary of changes** (Bug fix, feature, docs update, ...)

    Created a parse_intem function in openghg.standardise.emissions, this needs to

- [x]       read the NetCDF
- [x]       read metadata from the filename? Will it just be the whole DECC network for each file?
- [x]       create a metadata dictionary from the attributes (there aren't many), adding species, sites, network, domain, source = "intem"
- [x]       check the dataset using Emissions.validate_data  - (updated flux_mean to flux for validation)
- [x]       create a dictionary `{"data": parsed_dataset, "metadata": metadata_dict}
- [x]      return this dict

Next Steps :
    
- [x] Add an INTEM type to EmissionsTypes(Enum) in openghg.types._enum.py
- [x]     Make sure Emissions.read_file can use the parser to read the file
- [x]     Add some tests for the new data type, create a small (<= 50 kB if possible) NetCDF using ds.head(integer_that_makes_sense)
- [x]     Add adding INTEM files to the documentation
- [x]     Add to changelog

* **Please check if the PR fulfills these requirements**

- [x] Closes #797 (Replace xxxx with the Github issue number)
- [x] [Tests added and passed](https://docs.openghg.org/development/python_devel.html#testing) if fixing a bug or adding a new feature
- [x] All code checks passing - `black --line-length 110` run over code, `mypy` and `flake8` not showing any errors
- [x] Documentation and tutorials updated/added
- [x] Added an entry in the latest `CHANGELOG.md` file if fixing a bug or adding a new feature

### CODE REVIEW

**Comment by SutarPrasad** [comment]
Sample output:
{'n2o_intem_europe': {'data': <xarray.Dataset>
  Dimensions:             (time: 132, lon: 128, lat: 183, country: 2)
  Coordinates:
    * time                (time) datetime64[ns] 2012-01-01 ... 2022-12-01
    * lon                 (lon) float32 -14.12 -13.77 -13.42 ... 29.88 30.23 30.58
    * lat                 (lat) float32 36.47 36.7 36.94 ... 78.59 78.82 79.06
  Dimensions without coordinates: country
  Data variables:
      country_name        (country) |S8 ...
      flux                (lat, lon, time) float32 ...
      flux_std            (lat, lon, time) float32 ...
      flux_prior          (lat, lon, time) float32 ...
      flux_std_prior      (lat, lon, time) float32 ...
      country_post_mean   (country, time) float32 ...
      country_post_std    (country, time) float32 ...
      country_prior_mean  (country, time) float32 ...
      country_prior_std   (country, time) float32 ...
  Attributes:
      history:           Created on 2023-08-24 18:04:57 by A. J. Manning
      molecular_weight:  44.0
      author:            OpenGHG Cloud,
  'metadata': {'history': 'Created on 2023-08-24 18:04:57 by A. J. Manning',
   'molecular_weight': 44.0,
   'author': 'OpenGHG Cloud',
   'species': 'n2o',
   'domain': 'europe',
   'source': 'intem',
   'data_type': 'emissions',
   'processed': '2023-10-05 11:48:42.436004+00:00',
   'source_format': 'openghg',
   'start_date': '2012-01-01 00:00:00+00:00',
   'end_date': '2022-12-31 23:59:59+00:00',
   'max_longitude': 30.58,
   'min_longitude': -14.124,
   'max_latitude': 79.05701,
   'min_latitude': 36.469}}}

**Comment by SutarPrasad** [comment]
**SAMPLE OUTPUT**

{'n2o_intem_europe': {'data': <xarray.Dataset>
Dimensions:             (time: 132, lon: 128, lat: 183, country: 2)
Coordinates:
  * time                (time) datetime64[ns] 2012-01-01 ... 2022-12-01
  * lon                 (lon) float32 -14.12 -13.77 -13.42 ... 29.88 30.23 30.58
  * lat                 (lat) float32 36.47 36.7 36.94 ... 78.59 78.82 79.06
Dimensions without coordinates: country
Data variables:
    country_name        (country) |S8 ...
    flux                (lat, lon, time) float32 ...
    flux_std            (lat, lon, time) float32 ...
    flux_prior          (lat, lon, time) float32 ...
    flux_std_prior      (lat, lon, time) float32 ...
    country_post_mean   (country, time) float32 ...
    country_post_std    (country, time) float32 ...
    country_prior_mean  (country, time) float32 ...
    country_prior_std   (country, time) float32 ...
Attributes:
    history:           Created on 2023-08-24 18:04:57 by A. J. Manning
    molecular_weight:  44.0
    author:            OpenGHG Cloud, 'metadata': {'history': 'Created on 2023-08-24 18:04:57 by A. J. Manning', 'molecular_weight': 44.0, 'author': 'OpenGHG Cloud', 'species': 'n2o', 'domain': 'europe', 'source': 'intem', 'data_type': 'emissions', 'processed': '2023-10-05 12:56:59.198065+00:00', 'source_format': 'openghg', 'start_date': '2012-01-01 00:00:00+00:00', 'end_date': '2022-12-31 23:59:59+00:00', 'max_longitude': 30.58, 'min_longitude': -14.124, 'max_latitude': 79.05701, 'min_latitude': 36.469}}}

**Review by gareth-j — CHANGES_REQUESTED** [review]

File: openghg/standardise/emissions/__init__.py [hunk 1]
```diff
@@ -1 +1,2 @@
 from ._openghg import parse_openghg
+from ._openghg import parse_intem_emissions
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/__init__.py]: done

File: openghg/standardise/emissions/_openghg.py [hunk 2]
```diff
@@ -99,3 +99,100 @@ def parse_openghg(
...
+    database_version: Optional[str] = None,
+    model: Optional[str] = None,
+    period: Optional[Union[str, tuple]] = None,
+    chunks: Union[int, Dict, Literal["auto"], None] = None,
+    continuous: bool = True,
+) -> Dict:
+    """
+    Parse INTEM emissions data from the specified file.
+
+    Parameters:
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_openghg.py]: done, renamed parameters to Args

File: openghg/standardise/emissions/_openghg.py [hunk 3]
```diff
@@ -99,3 +99,100 @@ def parse_openghg(
...
+    - filepath (Path): Path to the '.nc' file containing INTEM emissions data.
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_openghg.py]: done. removed the dashes and indented by comparing to parse_openghg function for consistency

File: openghg/standardise/emissions/_openghg.py [hunk 4]
```diff
@@ -99,3 +99,100 @@ def parse_openghg(
...
+    - data_type (str): Type of data, default is 'emissions'.
+    - database (Optional[str]): Database name if applicable.
+    - database_version (Optional[str]): Version of the database if applicable.
+    - model (Optional[str]): Model name if applicable.
+    - period "needs explainer"
+    - chunks (Union[int, Dict, Literal["auto"], None]): Chunking configuration.
+    - continuous (bool): "needs explainer"
+
+    Returns:
+    - Dict: Parsed emissions data in dictionary format.
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_openghg.py]: done the same as above

File: openghg/standardise/emissions/_intem.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    model: Optional[str] = None,
+    period: Optional[Union[str, tuple]] = None,
+    chunks: Union[int, Dict, Literal["auto"], None] = None,
+    continuous: bool = True,
+) -> Dict:
+    """
+    Parse INTEM emissions data from the specified file.
+
+    Args:
+        filepath (Path): Path to the '.nc' file containing INTEM emissions data.
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:22]: We don't need to the types here as we've got the hinting above.

File: openghg/standardise/emissions/_intem.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
+from pathlib import Path
+from typing import Dict, Literal, Optional, Union
+
+
+def parse_intem(
+    filepath: Path,
+    species: str,
+    source: str = "intem",
+    domain: str = "europe",
+    data_type: str = "emissions",
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:10]: I don't think we need this as it's always going to be emissions?

File: openghg/standardise/emissions/_intem.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    database: Optional[str] = None,
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:11]: Do we have a database version for INTEM @rt17603 ?

File: openghg/standardise/emissions/_intem.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    Args:
+        filepath (Path): Path to the '.nc' file containing INTEM emissions data.
+        species (str): Name of species
+        source (str): Source of the emissions data, default is 'intem'.
+        domain (str): Geographic domain, default is 'europe'.
+        data_type (str): Type of data, default is 'emissions'.
+        database (Optional[str]): Database name if applicable.
+        database_version (Optional[str]): Version of the database if applicable.
+        model (Optional[str]): Model name if applicable.
+        period : "needs explainer"
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:30]: This needs updating

File: openghg/standardise/emissions/_intem.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+        chunks : (Union[int, Dict, Literal["auto"], None]): Chunking configuration.
+        continuous (bool): "needs explainer"
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:32]:
```suggestion
        continuous: Whether time stamps have to be continuous.
```

File: openghg/standardise/emissions/_intem.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+    Returns:
+        Dict: Parsed emissions data in dictionary format.
+    """
+    from openghg.util import timestamp_now
+    from openghg.store import infer_date_range
+    from openghg.store._emissions import Emissions
+    from xarray import open_dataset
+
+    emissions_dataset = open_dataset(filepath, chunks=chunks)
```
Comment (gareth-j) [inline openghg/standardise/emissions/_intem.py:42]:
```suggestion
    emissions_dataset = open_dataset(filepath).chunk(chunks)
```

Comment (gareth-j) [reply openghg/standardise/emissions/_intem.py:42]: xarray often doesn't chunk files correctly if you use the chunks argument to `open_dataset`, it's best to chunk after opening and chain the calls.

**Review by gareth-j — CHANGES_REQUESTED** [review]
Just a few updates needed, otherwise looks good!

File: openghg/standardise/emissions/_intem.py (hunk 9 above)
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py:32]: @rt17603 can you please advice on this?

File: openghg/standardise/emissions/_intem.py (hunk 7 above)
Comment (rt17603) [reply openghg/standardise/emissions/_intem.py:10]: INTEM isn't a database, it's an inversion model (run by Alistair and the met office).

For "emissions" we can expect them to come from different places (and can use different connected labels for this) e.g.
 - an output from a model run e.g. to simulate wetlands model (e.g. model)
 - bottom up inventory (e.g. database, database version etc.)
 - a posterior output from an inversion model (e.g. could use model again or could add a new input option)

File: openghg/standardise/emissions/_intem.py (hunk 9 above)
Comment (rt17603) [reply openghg/standardise/emissions/_intem.py:31]: With emissions/footprints/boundary conditions data in our standard format we often have to infer a time period associated with the time points. The "continuous" constraint comes in within the `openghg.store._infer_time.infer_date_range()` function where this is checking whether it has been possible to infer this detail e.g. https://github.com/openghg/openghg/blob/devel/openghg/store/_infer_time.py#L138.

If this is set to False this carries on without this piece of information (though I believe Gareth found a place this falls down https://github.com/openghg/openghg/issues/623 - the specific case got fixed but fixing the additional corner cases may have been deferred until they become clear). If this is set to True this raises a ValueError and exits.

**Review by rt17603 — CHANGES_REQUESTED** [review]
I assume the overall intention of adding this functionality in this PR is to allow the INTEM data to be added to the object store? If so, this is mostly there but there are one or two linking steps which need to be added.

Basically it should be possible to access your `parse_intem` function along the chain from  [`standardise_flux`](https://github.com/openghg/openghg/blob/devel/openghg/standardise/_standardise.py#L452). To do this I think you're just missing the following step:
 - Add `source_format` as an option for `standardise_flux` passing this onto `standardise()`.
     -  `source_format` is actually already an optional input in [`Emissions.read_file()`](https://github.com/openghg/openghg/blob/devel/openghg/store/_emissions.py#L50) which will be accessed from [`standardise()`](https://github.com/openghg/openghg/blob/devel/openghg/standardise/_standardise.py#L13).
     - If you then call `standardise_flux` with an input of `source_format="intem"`, as you've already matched to the file and function naming with your parse function, that *should* then find your `parse_intem` function, use it and go through the process of adding the data to the object store.

I would also suggest adding in another test to check that works as expected (i.e. using `standardise_flux` with `source_format="intem"` and passing your "ch4_intem.nc" file). See "tests/standardise/test_standardise.py" for examples e.g. something like [`test_standardise_flux`](https://github.com/openghg/openghg/blob/devel/tests/standardise/test_standardise.py#L306).

File: openghg/standardise/emissions/_intem.py [hunk 11]
```diff
@@ -0,0 +1,99 @@
+from pathlib import Path
+from typing import Dict, Literal, Optional, Union
+
+
+def parse_intem(
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py:5]: Agreed, Yes, later we can input additional details.

File: openghg/standardise/emissions/_intem.py [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+    metadata["min_longitude"] = round(float(emissions_dataset["lon"].min()), 5)
+    metadata["max_latitude"] = round(float(emissions_dataset["lat"].max()), 5)
+    metadata["min_latitude"] = round(float(emissions_dataset["lat"].min()), 5)
+
+    key = "_".join((species, source, domain))
+
+    emissions_dataset = emissions_dataset.rename_vars({"flux_mean": "flux"})
+
+    # Dataset validation
+    Emissions.validate_data(emissions_dataset)
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py:92]: done

File: openghg/standardise/emissions/_intem.py [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+    """
+    Parse INTEM emissions data from the specified file.
+
+    Args:
+        filepath: Path to the '.nc' file containing INTEM emissions data.
+        species: Name of species
+        data_type: Type of data, default is 'emissions'.
+        source: Source of the emissions data, default is 'intem'.
+        domain: Geographic domain, default is 'europe'.
+        database: Database name if applicable.
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py]: Standardise flux has database and database_version, and we also supply those option_keywords in emissions.read_file(). Thus to maintain cohesion I have added position arguments inside parse_intem but avoided its inclusion inside the metadata. 
@gareth-j and @rt17603 Do let me know your thoughts and if there is any other way to apply the desired changes.

File: openghg/standardise/emissions/_intem.py [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+        database_version: Version of the database if applicable.
+        model: Model name if applicable.
+        period: The time period for which data is to be parsed.
+        chunks: (Union[int, Dict, Literal["auto"], None]): Chunking configuration.
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py]: done

File: openghg/standardise/emissions/_intem.py [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+    period: Optional[Union[str, tuple]] = None,
+    high_time_resolution: Optional[bool] = False,
+    chunks: Union[int, Dict, Literal["auto"], None] = None,
+    continuous: bool = True,
+) -> Dict:
+    """
+    Parse INTEM emissions data from the specified file.
+
+    Args:
+        filepath: Path to the '.nc' file containing INTEM emissions data.
```
[the remark this answers is not in the document]
Comment (SutarPrasad) [reply openghg/standardise/emissions/_intem.py:22]: done

**Comment by SutarPrasad** [comment]
Added **kwargs to avoid getting stuck into the error of not passing "database" and "database_version" variables, usually passed in the standardise_flux function.

**Review by gareth-j — APPROVED** [review]
Looks good! Thanks for making those changes.

**Pull request merged**